### PR TITLE
Move the JSON parser from evalengine

### PR DESCRIPTION
**Pull request by dbussink**
This moves the JSON evalengine implementation so we can share it more across different components.

Specifically vreplication wants to also use the same JSON parser so it can handle dealing with the MySQL types inside JSON objects better.

It also implements the same parsing for the MySQL binary JSON format and a serialization to using SQL style functions to limit typing loss as much as possible.

There's still two types that cause problems. When using `TIMESTAMP` or `BIT` types inside a JSON document it's impossible to create an SQL format compatible with those types since there's no way to create literals for those two types.

## Related Issue(s)

This helps with a better fix for #12731 that handles more types correctly. 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by vmg — APPROVED** [review]
This is just moving/renaming, nothing to review :ok_hand:

File: go/mysql/float.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package mysql
+
```
Comment (frouioui) [inline go/mysql/float.go:2]: We need to add a header here

File: go/mysql/json/mysql_parser_test.go [hunk 2]
```diff
@@ -0,0 +1,430 @@
+package json
+
```
Comment (frouioui) [inline go/mysql/json/mysql_parser_test.go:2]: same here

**Review by frouioui — APPROVED** [review]
LGTM

**Comment by dbussink** [comment]
Looks like this was merged but the PR not closed 😞

**Pull request closed without merging**